### PR TITLE
[SMALLFIX] [branch-1.3] Set timeout for transport open

### PR DESCRIPTION
**Pull request by ifcharming**
to avoid client hanging when authentication types do not match between client and servers.

### CODE REVIEW

**Comment by ifcharming** [comment]
@calvinjia

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 1]
```diff
@@ -160,10 +167,13 @@ public synchronized void connect() throws IOException, ConnectionFailedException
     disconnect();
     Preconditions.checkState(!mClosed, "Client is closed, will not try to connect.");
 
+    int socketTimeoutMs = Configuration.getInt(
+        PropertyKey.SECURITY_AUTHENTICATION_SOCKET_TIMEOUT_MS);
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/AbstractClient.java]: Would it be possible to format with the new line starting after the `=`?

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 2]
```diff
@@ -160,10 +167,13 @@ public synchronized void connect() throws IOException, ConnectionFailedException
...
     int maxConnectsTry = Configuration.getInt(PropertyKey.MASTER_RETRY);
     final int BASE_SLEEP_MS = 50;
     RetryPolicy retry =
         new ExponentialBackoffRetry(BASE_SLEEP_MS, Constants.SECOND_MS, maxConnectsTry);
+    ExecutorService executor = Executors.newCachedThreadPool();
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/AbstractClient.java]: Each client will create its own thread pool? Also, we should make the thread daemon threads and shutdown the executor when possible.

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 3]
```diff
@@ -173,7 +183,23 @@ public synchronized void connect() throws IOException, ConnectionFailedException
           new TBinaryProtocol(mTransportProvider.getClientTransport(mAddress));
       mProtocol = new TMultiplexedProtocol(binaryProtocol, getServiceName());
       try {
-        mProtocol.getTransport().open();
+        // Set a timeout for transport.open() in case it hangs when the authentication type does not
```
Comment (peisun1115) [inline core/common/src/main/java/alluxio/AbstractClient.java]: Can you do this for ThriftClientPool.java as well?

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 4]
```diff
@@ -173,7 +183,23 @@ public synchronized void connect() throws IOException, ConnectionFailedException
...
+        try {
+          openTask.get((long) socketTimeoutMs, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+          String message = "Timeout to open the thrift transport. This exception may be caused by "
+              + "incorrect network configuration. Please check whether the authentication types "
+              + "match between client and server. Note that NOSASL client is not able to connect "
+              + "to servers with SIMPLE security mode.";
+          throw new IOException(message, e);
+        } finally {
+          openTask.cancel(true);
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/AbstractClient.java]: Does it make sense to cancel it even on success?

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 5]
```diff
@@ -191,7 +217,7 @@ public synchronized void connect() throws IOException, ConnectionFailedException
           throw new IOException(message, e);
         }
         throw e;
-      } catch (TTransportException e) {
+      } catch (Exception e) {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/AbstractClient.java]: Is it intended to catch the previously thrown `IOException`?

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 6]
```diff
@@ -350,4 +376,23 @@ protected synchronized InetSocketAddress getAddress() {
     }
     throw new IOException("Failed after " + retry + " retries.");
   }
+
+  private class TransportOpenCallable implements Callable<Void> {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/AbstractClient.java]: Would it be possible to implement this as an anonymous class? It seems fairly one-time use.

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 7]
```diff
@@ -173,7 +183,23 @@ public synchronized void connect() throws IOException, ConnectionFailedException
...
       mProtocol = new TMultiplexedProtocol(binaryProtocol, getServiceName());
       try {
-        mProtocol.getTransport().open();
+        // Set a timeout for transport.open() in case it hangs when the authentication type does not
+        // match between Alluxio client and servers.
+        FutureTask<Void> openTask = new FutureTask<>(
+            new TransportOpenCallable(mProtocol.getTransport()));
+        executor.submit(openTask);
+        try {
+          openTask.get((long) socketTimeoutMs, TimeUnit.MILLISECONDS);
```
Comment (peisun1115) [inline core/common/src/main/java/alluxio/AbstractClient.java]: This doesn't really timeout the for connection open request. Why do you still need this given we have a timeout in TransportProvierUtils#createThriftSocket which times out for creating a socket.

File: core/common/src/main/java/alluxio/AbstractClient.java (hunk 4 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/AbstractClient.java]: hmm, does this cancel work? AFAIK, future.cancel only works if the underlying task can be cancelled (for example, sleep).

**Comment by ifcharming** [comment]
@calvinjia  Discussed with @peisun1115 offline...
Closing this PR because actually the socket timeout `alluxio.security.authentication.socket.timeout.ms` is indeed respected in `transport.open`.
The client seems to hang just because the timeout has default value of 10min, and retry for 30 times.... We should instead consider downgrade the socket timeout or do not retry on timeout.

**Pull request closed without merging**